### Tarea Random Forests
#### Daniel Espinosa 136981


El objetivo de este trabajo es comparar el desempeño de los modelos **Random Forest** y **Árbol de decisión**.

Se espera ver que Random Forest tenga mejor desempeño ya que es un modelo más robusto pues es es un modelo en ensamble compuesto por varios árboles de decisión.

El dataset que se utiliza es el [
Breast Cancer Wisconsin (Diagnostic) Data Set](https://archive.ics.uci.edu/ml/datasets/Breast+Cancer+Wisconsin+%28Diagnostic%29) que contiene varias variables explicativas numéricas derivadas de muestras de tejido de mama de diversas pacientes y la variable target es **Diagnosis**, si el tejido fue maligno (cáncer) o benigno (no cáncer).

Los modelos se entrenarán para clasificar el diagnóstico de los tejidos.

Las variables del dataset son las siguientes:

1) ID number

2) Diagnosis (M = malignant, B = benign) 

Ten real-valued features are computed for each cell nucleus: 

a) radius (mean of distances from center to points on the perimeter) 

b) texture (standard deviation of gray-scale values) 

c) perimeter 

d) area 

e) smoothness (local variation in radius lengths) 

f) compactness (perimeter^2 / area - 1.0) 

g) concavity (severity of concave portions of the contour) 

h) concave points (number of concave portions of the contour) 

i) symmetry 

j) fractal dimension ("coastline approximation" - 1)

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier

from sklearn.metrics import confusion_matrix

In [3]:
df = pd.read_csv("wdbc.csv",header=None, na_values = "?")
df.describe()

0           2           3           4            5   \
count  5.690000e+02  569.000000  569.000000  569.000000   569.000000   
mean   3.037183e+07   14.127292   19.289649   91.969033   654.889104   
std    1.250206e+08    3.524049    4.301036   24.298981   351.914129   
min    8.670000e+03    6.981000    9.710000   43.790000   143.500000   
25%    8.692180e+05   11.700000   16.170000   75.170000   420.300000   
50%    9.060240e+05   13.370000   18.840000   86.240000   551.100000   
75%    8.813129e+06   15.780000   21.800000  104.100000   782.700000   
max    9.113205e+08   28.110000   39.280000  188.500000  2501.000000   

               6           7           8           9           10     ...      \
count  569.000000  569.000000  569.000000  569.000000  569.000000     ...       
mean     0.096360    0.104341    0.088799    0.048919    0.181162     ...       
std      0.014064    0.052813    0.079720    0.038803    0.027414     ...       
min      0.052630    0.019380    0.000000    0.000000    0.106000     ...       
25%      0.086370    0.064920    0.029560    0.020310    0.161900     ...       
50%      0.095870    0.092630    0.061540    0.033500    0.179200     ...       
75%      0.105300    0.130400    0.130700    0.074000    0.195700     ...       
max      0.163400    0.345400    0.426800    0.201200    0.304000     ...       

               22          23          24           25          26  \
count  569.000000  569.000000  569.000000   569.000000  569.000000   
mean    16.269190   25.677223  107.261213   880.583128    0.132369   
std      4.833242    6.146258   33.602542   569.356993    0.022832   
min      7.930000   12.020000   50.410000   185.200000    0.071170   
25%     13.010000   21.080000   84.110000   515.300000    0.116600   
50%     14.970000   25.410000   97.660000   686.500000    0.131300   
75%     18.790000   29.720000  125.400000  1084.000000    0.146000   
max     36.040000   49.540000  251.200000  4254.000000    0.222600   

               27          28          29          30          31  
count  569.000000  569.000000  569.000000  569.000000  569.000000  
mean     0.254265    0.272188    0.114606    0.290076    0.083946  
std      0.157336    0.208624    0.065732    0.061867    0.018061  
min      0.027290    0.000000    0.000000    0.156500    0.055040  
25%      0.147200    0.114500    0.064930    0.250400    0.071460  
50%      0.211900    0.226700    0.099930    0.282200    0.080040  
75%      0.339100    0.382900    0.161400    0.317900    0.092080  
max      1.058000    1.252000    0.291000    0.663800    0.207500  

[8 rows x 31 columns]

Hay variables que no son relevantes como el ID. A continuación se forma la matriz de los datos como se utilizarán.

In [25]:
data = df.as_matrix()

data_final = np.column_stack((data[:,2:],data[:,1]))

np.random.seed(923)
X_train, X_test, Y_train, Y_test = train_test_split(data_final[:,:-1],data_final[:,-1], train_size=0.75)

### Desempeño del árbol de decisión

Árbol de decisión con min_samples_split = 20 para que el árbol no sea tan profundo y generalice mejor.

In [46]:
clf = DecisionTreeClassifier(min_samples_split = 20)
clf = clf.fit(X_train, Y_train)
Y_test_prediction = clf.predict(X_test)
cm = confusion_matrix(Y_test, Y_test_prediction)
tn, fp, fn, tp = confusion_matrix(Y_test, Y_test_prediction).ravel()
cm

array([[80,  3],
       [ 5, 55]])

In [44]:
#Accuracy (num buenos/total)
accuracy = float(tp+tn) / float(tp + tn + fp + fn)
print("Accuracy: " + str(accuracy))
#Precision

precision = float(tp)/float(tp + fp)
print("Precision: " + str(precision))
#Recall

recall = float(tp)/float(tp + fn)
print("Recall: " + str(recall))

#F1

f1 = f1 = (2*(precision*recall))/(precision+recall)
print("F1 score: " + str(f1))

Accuracy: 0.937062937063
Precision: 0.947368421053
Recall: 0.9
F1 score: 0.923076923077


### Desempeño del Random Forest

Random Forest con 15 árboles y también min_samples_split = 20.

In [47]:
clf = RandomForestClassifier(n_estimators = 15 ,min_samples_split = 20)
clf = clf.fit(X_train, Y_train)
Y_test_prediction = clf.predict(X_test)
cm = confusion_matrix(Y_test, Y_test_prediction)
tn, fp, fn, tp = confusion_matrix(Y_test, Y_test_prediction).ravel()
cm

array([[81,  2],
       [ 2, 58]])

In [48]:
#Accuracy (num buenos/total)
accuracy = float(tp+tn) / float(tp + tn + fp + fn)
print("Accuracy: " + str(accuracy))
#Precision

precision = float(tp)/float(tp + fp)
print("Precision: " + str(precision))
#Recall

recall = float(tp)/float(tp + fn)
print("Recall: " + str(recall))

#F1

f1 = f1 = (2*(precision*recall))/(precision+recall)
print("F1 score: " + str(f1))

Accuracy: 0.972027972028
Precision: 0.966666666667
Recall: 0.966666666667
F1 score: 0.966666666667


### Comparación del desempeño de los modelos

|Métrica|**Árbol de decisión**|**Random Forest**|
|---|---|---|
|**Accuracy**|0.937062937063|0.972027972028|
|**Precision**|0.947368421053|0.966666666667|
|**Recall**|0.9|0.966666666667|
|**F1 Score**|0.923076923077|0.966666666667|

### Conclusiones

Aunque el árbol de decisión obtuvo resultados buenos, el modelo Random Forest obtuvo resultados todavía mejores y por un márgen considerable. Random Forest, al ser un meta modelo de aprendizaje en ensamble y en este caso hacer bootstrap, combina la predicción de muchos árboles construídos aleatoriamente y entrenados con diferentes muestras con reemplazo del set de entrenamiento, lo que ayuda a que controle el *overfitting*.

Es importante tomar en cuenta que muchas de las técnicas de aprendizaje de máquina se pueden combinar en diferentes maneras para formar modelos más robustos y que tengan mejor desempeño al clasificar o hacer regresiones sobre diferentes tipos de datos.